app/models/host/base.rb

```diff
@@ -214,14 +214,18 @@ def set_comment(parser)
       desc 'Note that available facts depend on what facts have been uploaded to Foreman,
            typical sources are Puppet facter, subscription manager etc.
            The facts can be out of date, this macro only provides access to the value stored in the database.'
+      optional :fact_names, Array, desc: 'A list of fact names to return. If empty all facts are returned'
       returns Hash, desc: 'A hash of facts, keys are fact names, values are fact values'
       example '@host.facts # => { "hardwareisa"=>"x86_64", "kernel"=>"Linux", "virtual"=>"physical", ... }', desc: 'Getting all host facts'
       example '@host.facts["uptime"] # => "30 days"', desc: 'Getting specific fact value, +uptime+ in this case'
       aliases :facts
     end
-    def facts_hash
+    def facts_hash(fact_names = [], *args)
+      filtered_names = [fact_names, *args].flatten.compact
       hash = {}
-      fact_values.includes(:fact_name).collect do |fact|
+      query = fact_values.includes(:fact_name)
+      query = query.where(fact_names: {name: filtered_names}) if filtered_names.present?
+      query.each do |fact|
         hash[fact.fact_name.name] = fact.value
       end
       hash
```

Comment:
Just a quick suggestion: I think this comes down to:
```suggestion
      query.to_h { |fact| [fact.fact_name.name, fact.value] }
```

But that also makes me think: do you need to retrieve all columns? I suspect you can use `pluck()`. Something like this probably works:
```suggestion
      query.pluck('fact_name.name', :value)
```

Then also drop `hash = {}`.